### ISSUE TITLE
Question about implementation of log softmax

### ISSUE THREAD

**Issue by linzhiqiu** [azshue/TPT#5]
I am curious why you used 

`logits = outputs - outputs.logsumexp(dim=-1, keepdim=True)` 
instead of 
` logits = outputs.log_softmax(dim=1) [N, 1000]`

and similarly
`avg_logits = logits.logsumexp(dim=0) - np.log(logits.shape[0])`
instead of 
`avg_logits = logits.mean(0) [1, 1000]`

Thanks!

**Comment by azshue** [comment]
Hi @linzhiqiu,

Thank you for your interest in our work and implementation.

For calculating the averaged entropy, we follow the [implementation of MEMO](https://github.com/zhangmarvin/memo/blob/main/imagenet-exps/test_calls/test_adapt.py#L40). Using `logsumexp` is [a trick for numerical stability](https://en.wikipedia.org/wiki/LogSumExp#log-sum-exp_trick_for_log-domain_calculations), but it could be that PyTorch's implementation of `log_softmax` has accounted for this issue, so it may be fine using either. 
(I don't recall if I've had problems using the alternative implementation. It's likely that I haven't used  the alternatives.)

**Comment by linzhiqiu** [comment]
Okay that makes a lot of sense -- I recently came across this issue in my other project and was worried that torch's log_softmax doesn't use this cool trick.

Thanks! Very impressive work and we are considering following up!

**Comment by scyonggg** [comment]
Just wanted to leave a note here for future researchers or developers who might try to refactor the `avg_entropy` function.

Ref: https://github.com/azshue/TPT/blob/63ecbace79694205d7884e63fdc3137a200f0b0e/tpt_classification.py#L46-L51

## Summary
I attempted to replace the manual log-probability calculation (`outputs - outputs.logsumexp(...)`) with PyTorch's `F.log_softmax()` to improve numerical stability.

I confirmed that PyTorch's native implementation already includes the "Log-Sum-Exp trick" for stability, as seen in the source code here:
https://github.com/pytorch/pytorch/blob/0e227bfc4f0d7b8054f63fb9d16ce3485182a771/aten/src/ATen/native/ReduceOps.cpp#L1490-L1505

However, contrary to expectations, I observed a **performance drop of approximately 4%p of the Out-of-Distribution Generalization task** with this 'numerically stable' version.


## Details
I tested the following code, which uses the fused `log_softmax` kernel and should theoretically provide cleaner gradients:

```python
def avg_entropy(outputs: torch.Tensor) -> torch.Tensor:
    # Compute log probabilities [N, C]
    log_probs = torch.log_softmax(outputs, dim=-1)

    # Average probabilities using logsumexp for numerical stability
    # log(mean(exp(x))) = logsumexp(x) - log(N)
    avg_log_probs = torch.logsumexp(log_probs, dim=0) - torch.log(
        torch.tensor(outputs.shape[0], dtype=outputs.dtype, device=outputs.device)
    )

    # Clamp for numerical stability
    min_real = torch.finfo(avg_log_probs.dtype).min
    avg_log_probs = torch.clamp(avg_log_probs, min=min_real)

    # Compute entropy: H = -sum(p * log(p))
    avg_probs = torch.exp(avg_log_probs)
    entropy = -(avg_probs * avg_log_probs).sum(dim=-1)

    return entropy
```

## Conclusion
While the gradients from `log_softmax` are technically "cleaner" (less rounding error), the model seems to perform better with the original implementation ().
I haven't pinpointed the exact reason yet—whether the model benefits from the specific noise profile of the manual calculation or if there's another underlying factor. I would be happy to hear if anyone has insights on why this discrepancy occurs.
